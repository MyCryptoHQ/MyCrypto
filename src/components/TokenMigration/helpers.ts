import { getAssetByUUID } from '@services/Store/Asset';
import { ITokenMigrationConfig, ITxConfig, ITxObject, StoreAccount } from '@types';
import { hexToString, hexWeiToString } from '@utils';

export const makeTokenMigrationTxConfig = (
  rawTransaction: ITxObject,
  account: StoreAccount,
  amount: string
) => (tokenMigrationConfig: ITokenMigrationConfig): ITxConfig => {
  const { gasPrice, gasLimit, nonce, data, value } = rawTransaction;
  const { address, network } = account;
  const baseAsset = getAssetByUUID(account.assets)(network.baseAsset)!;
  const asset = getAssetByUUID(account.assets)(tokenMigrationConfig.fromAssetUuid)!;
  const txConfig: ITxConfig = {
    from: address,
    receiverAddress: account.address,
    senderAccount: account,
    network,
    asset,
    baseAsset,
    amount,
    gasPrice: hexToString(gasPrice),
    gasLimit: hexToString(gasLimit),
    value: hexWeiToString(value),
    nonce: hexToString(nonce),
    data,
    rawTransaction
  };

  return txConfig;
};
