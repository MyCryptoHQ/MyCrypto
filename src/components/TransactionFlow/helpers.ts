import isEmpty from 'ramda/src/isEmpty';
import mergeDeepWith from 'ramda/src/mergeDeepWith';
import pick from 'ramda/src/pick';
import { ValuesType } from 'utility-types';

import { bigNumGasPriceToViewableGwei } from '@services/EthService/utils';
import { getAccountBalance, getStoreAccount } from '@services/Store';
import { ITxConfig, StoreAccount } from '@types';

import { ISender } from './types';

type FieldValue = ValuesType<ISender>;

const preferValueFromSender = (l: FieldValue, r: FieldValue): FieldValue => (isEmpty(r) ? l : r);

export const constructSenderFromTxConfig = (
  txConfig: ITxConfig,
  accounts: StoreAccount[]
): ISender => {
  const { network, senderAccount, from } = txConfig;
  const defaultSender: ISender = {
    address: from,
    assets: [],
    network
  };

  const sender: ISender = mergeDeepWith(
    preferValueFromSender,
    defaultSender,
    pick(['address', 'assets', 'network'], {
      ...senderAccount
    })
  );

  // if account exists in store add it to sender
  const account = getStoreAccount(accounts)(sender.address, sender.network.id);
  if (account) {
    sender.account = account;
    sender.accountBalance = getAccountBalance(senderAccount);
  }

  return sender;
};

// replacement gas price must be at least 10% higher than the replaced tx's gas price
export const calculateReplacementGasPrice = (txConfig: ITxConfig, fastGasPrice: number) =>
  Math.max(fastGasPrice, parseFloat(bigNumGasPriceToViewableGwei(txConfig.gasPrice)) * 1.101);
