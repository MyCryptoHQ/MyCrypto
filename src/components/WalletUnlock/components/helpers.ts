import { DPath } from '@types';

import { TableAccountDisplay } from './DeterministicAccountTable';

export const sortAccounts = (
  accounts: TableAccountDisplay[],
  displayEmptyAddresses: boolean,
  selectedDPath: DPath
) => {
  const selectedAccounts = accounts
    .filter(({ isSelected }) => isSelected)
    .sort((a, b) => (a.balance?.isGreaterThan(b.balance!) ? -1 : 1));
  const deselectedAccounts = accounts
    .filter(
      ({ isSelected, pathItem: { baseDPath } }) =>
        !isSelected && baseDPath.value === selectedDPath.value
    )
    .sort((a, b) => (a.pathItem.index < b.pathItem.index ? -1 : 1));
  return displayEmptyAddresses ? [...selectedAccounts, ...deselectedAccounts] : selectedAccounts;
};
