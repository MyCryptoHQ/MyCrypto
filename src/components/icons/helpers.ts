export type IconSize = 'xs' | 'sm' | 'base' | 'lg' | 'xl';

export const deriveIconSize = (size: IconSize = 'base') => {
  switch (size) {
    case 'xs':
      return '8';
    case 'sm':
      return '12';
    default:
    case 'base':
      return '16';
    case 'lg':
      return '20';
    case 'xl':
      return '24';
  }
};
