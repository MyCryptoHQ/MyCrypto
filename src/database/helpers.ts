import { FlowTransducer, GenObject } from './types';

export const toArray = (object: any): any[] => Object.values(object);

export const toObject = <T extends TObject>(keyInElem: keyof T) => (
  acc: GenObject<T>,
  elem: T
): GenObject<T> => ({
  ...acc,
  [elem[keyInElem] as any]: elem
});

export const add: FlowTransducer = (key) => (fn) => (data) => (store) => {
  return {
    ...store,
    [key]: fn(data, store)
  };
};
