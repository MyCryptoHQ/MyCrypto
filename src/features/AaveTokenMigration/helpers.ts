import { DEFAULT_ASSET_DECIMAL, DEFAULT_NETWORK_CHAINID } from '@config';
import { formatApproveTx } from '@helpers';
import { AaveMigrator } from '@services/EthService/contracts';
import { ITokenMigrationFormFull, ITxData, ITxObject, ITxToAddress } from '@types';
import { inputGasPriceToHex, inputValueToHex, toWei } from '@utils';

import { MIGRATION_CONTRACT } from './config';

export const createApproveTx = (
  payload: ITokenMigrationFormFull
): Omit<ITxObject, 'nonce' | 'gasLimit'> =>
  formatApproveTx({
    contractAddress: payload.asset.contractAddress as ITxToAddress,
    baseTokenAmount: toWei(payload.amount, DEFAULT_ASSET_DECIMAL),
    fromAddress: payload.account.address,
    spenderAddress: MIGRATION_CONTRACT,
    chainId: DEFAULT_NETWORK_CHAINID,
    hexGasPrice: inputGasPriceToHex(payload.gasPrice)
  });

export const createMigrationTx = (
  payload: ITokenMigrationFormFull
): Omit<ITxObject, 'nonce' | 'gasLimit'> => {
  const data = AaveMigrator.migrateFromLEND.encodeInput({
    amount: toWei(payload.amount, DEFAULT_ASSET_DECIMAL)
  });
  return {
    from: payload.account.address,
    to: MIGRATION_CONTRACT,
    value: inputValueToHex('0'),
    data: data as ITxData,
    gasPrice: inputGasPriceToHex(payload.gasPrice),
    chainId: DEFAULT_NETWORK_CHAINID
  };
};
