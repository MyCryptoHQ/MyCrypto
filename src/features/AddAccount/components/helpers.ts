import { DerivationPath as DPath } from '@mycrypto/wallets';

import { bigify, hasBalance } from '@utils';

import { TableAccountDisplay } from './HDWTable';

export const sortAccounts = (
  accounts: TableAccountDisplay[],
  displayEmptyAddresses: boolean,
  selectedDPath: DPath
) => {
  const selectedAccounts = accounts
    .filter(({ isSelected, balance }) => (isSelected && balance) || hasBalance(balance))
    .sort((a, b) => (bigify(a.balance!).isGreaterThan(b.balance!) ? -1 : 1));
  const deselectedAccounts = accounts
    .filter(
      ({ isSelected, pathItem: { baseDPath } }) =>
        !isSelected && baseDPath.path === selectedDPath.path
    )
    .sort((a, b) => (a.pathItem.index < b.pathItem.index ? -1 : 1));
  return displayEmptyAddresses ? [...selectedAccounts, ...deselectedAccounts] : selectedAccounts;
};

export const calculateDPathOffset = (accounts: TableAccountDisplay[], selectedDPath: DPath) =>
  Math.max(
    ...accounts
      .filter((acc) => acc.pathItem.baseDPath.path === selectedDPath.path)
      .map(({ pathItem: { index } }) => index)
  ) + 1; // Start scanning from the next index
