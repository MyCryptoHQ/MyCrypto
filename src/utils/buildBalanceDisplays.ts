import { StoreAsset, StoreAccount, ISettings, Balance, Asset, TUuid, BalanceAccount } from '@types';
import { translateRaw } from '@translations';

import { weiToFloat, convertToFiatFromAsset } from './convert';

export const buildBalances = (
  totals: (selectedAccounts?: StoreAccount[]) => StoreAsset[],
  accounts: StoreAccount[],
  settings: ISettings,
  getAssetRate: (asset: Asset) => number | undefined,
  filter: (excludedAssetUuids: TUuid[]) => (asset: StoreAsset) => boolean
): Balance[] =>
  totals(accounts)
    .filter(filter(settings.excludedAssets))
    .map((asset: StoreAsset) => {
      const exchangeRate = getAssetRate(asset);
      return {
        id: `${asset.name}-${asset.ticker}`,
        name: asset.name || translateRaw('WALLET_BREAKDOWN_UNKNOWN'),
        ticker: asset.ticker,
        uuid: asset.uuid,
        amount: weiToFloat(asset.balance, asset.decimal),
        fiatValue: convertToFiatFromAsset(asset, exchangeRate),
        exchangeRate,
        accounts: accounts.reduce((acc, currAccount) => {
          const matchingAccAssets = currAccount.assets.filter(
            (accAsset) => accAsset.uuid === asset.uuid
          );
          if (matchingAccAssets.length) {
            return [
              ...acc,
              ...matchingAccAssets.map((accAsset) => ({
                address: currAccount.address,
                ticker: accAsset.ticker,
                amount: weiToFloat(accAsset.balance, accAsset.decimal),
                fiatValue: convertToFiatFromAsset(accAsset, exchangeRate),
                label: currAccount.label
              }))
            ];
          }
          return acc;
        }, [] as BalanceAccount[])
      };
    })
    .sort((a, b) => b.fiatValue - a.fiatValue);

export const buildTotalFiatValue = (balances: Balance[]) =>
  balances.reduce((sum, asset) => {
    return sum + asset.fiatValue;
  }, 0);
