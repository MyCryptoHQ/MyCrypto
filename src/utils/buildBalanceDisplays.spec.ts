import { StoreAccount, Asset, TUuid, StoreAsset } from '@types';
import { fAccounts, fStoreAssets, fSettings } from '@fixtures';
import { getTotalByAsset } from '@services/Store/Asset/helpers';
import { buildBalances, buildTotalFiatValue } from './buildBalanceDisplays';

const totals = (_: StoreAccount[]) =>
  Object.values(getTotalByAsset([...fStoreAssets, ...fStoreAssets]));
const getAssetRate = (_: Asset) => 1;
const filterTrue = (_: TUuid[]) => (__: StoreAsset) => true;
const filterFalse = (_: TUuid[]) => (__: StoreAsset) => false;

describe('buildBalances', () => {
  it('builds balances', () => {
    const actual = buildBalances(totals, fAccounts, fSettings, getAssetRate, filterTrue);
    expect(actual).toMatchSnapshot();
  });
});

describe('buildTotalFiatValue', () => {
  it('builds total fiat values ', () => {
    const balances = buildBalances(totals, fAccounts, fSettings, getAssetRate, filterFalse);
    const actual = buildTotalFiatValue(balances);
    expect(actual).toMatchSnapshot();
  });
});
