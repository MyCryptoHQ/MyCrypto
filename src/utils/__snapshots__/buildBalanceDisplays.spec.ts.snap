// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildBalances builds balances 1`] = `
Array [
  Object {
    "accounts": Array [],
    "amount": 4.2e-17,
    "exchangeRate": 1,
    "fiatValue": 4.2e-17,
    "id": "RopstenETH-RopstenETH",
    "name": "RopstenETH",
    "ticker": "RopstenETH",
    "uuid": "01f2d4ec-c263-6ba8-de38-01d66c86f309",
  },
  Object {
    "accounts": Array [],
    "amount": 2e-18,
    "exchangeRate": 1,
    "fiatValue": 2e-18,
    "id": "WrappedETH-WETH",
    "name": "WrappedETH",
    "ticker": "WETH",
    "uuid": "10e14757-78bb-4bb2-a17a-8333830f6698",
  },
  Object {
    "accounts": Array [],
    "amount": 2e-18,
    "exchangeRate": 1,
    "fiatValue": 2e-18,
    "id": "Ether-ETH",
    "name": "Ether",
    "ticker": "ETH",
    "uuid": "f7e30bbe-08e2-41ce-9231-5236e6aab702",
  },
  Object {
    "accounts": Array [],
    "amount": 2e-18,
    "exchangeRate": 1,
    "fiatValue": 2e-18,
    "id": "GoerliETH-GoerliETH",
    "name": "GoerliETH",
    "ticker": "GoerliETH",
    "uuid": "12d3cbf2-de3a-4050-a0c6-521592e4b85a",
  },
]
`;

exports[`buildTotalFiatValue builds total fiat values  1`] = `0`;
